### PR TITLE
[tech] Ids instead of indexes on stop time ref v2

### PR DESCRIPTION
**Pull request by patochectp**

### CODE REVIEW

File: src/add_prefix.rs [hunk 1]
```diff
@@ -52,6 +53,21 @@ where
     }
 }
 
+fn add_prefix_on_vehicle_journey_ids(
+    vehicle_journey_ids: &HashMap<(String, u32), String>,
```
Comment (woshilapin) [inline src/add_prefix.rs:57]: Is it possible to take a `&mut` instead of reallocating entirely a new `HashMap` ?

File: src/model.rs [hunk 2]
```diff
@@ -122,11 +122,14 @@ pub struct Collections {
     pub geometries: CollectionWithId<Geometry>,
     pub admin_stations: Collection<AdminStation>,
     #[serde(skip)]
-    pub stop_time_headsigns: HashMap<(Idx<VehicleJourney>, u32), String>,
+    //HashMap<vehicle_journey_id, stop_sequence, String>,
```
Comment (woshilapin) [inline src/model.rs:125]:
```suggestion
    //HashMap<(vehicle_journey_id, stop_sequence), headsign>,
```

File: src/ntfs/filter.rs [hunk 3]
```diff
@@ -144,21 +144,14 @@ fn filter_from_idxset<T: Id<T>>(
...
+) -> HashMap<(String, u32), String> {
     let mut updated_attributes_map = HashMap::new();
-    for (&(old_vj_idx, sequence), attribute) in attributes_map {
-        if let Some(new_vj_idx) = old_vj_idx_to_vj_id
-            .get(&old_vj_idx)
-            .and_then(|vj_id| vehicle_journeys.get_idx(vj_id))
-        {
-            updated_attributes_map.insert((new_vj_idx, sequence), attribute.clone());
+    for ((vj_id, sequence), attribute) in attributes_map {
+        if vehicle_journeys.get(&vj_id).is_some() {
```
Comment (woshilapin) [inline src/ntfs/filter.rs:153]:
```suggestion
        if vehicle_journeys.contains_key(&vj_id) {
```

File: src/ntfs/filter.rs [hunk 4]
```diff
@@ -144,21 +144,14 @@ fn filter_from_idxset<T: Id<T>>(
     });
 }
 
-fn updated_stop_time_attributes<T>(
+fn updated_stop_time_attributes(
```
Comment (woshilapin) [inline src/ntfs/filter.rs:147]: I have the feeling that this entire function could be rewritten in 1 line.

```rust
fn updated_stop_time_attributes(
    vehicle_journeys: &CollectionWithId<VehicleJourney>,
    attributes_map: HashMap<(String, u32), String>,
) -> HashMap<(String, u32), String> {
    attributes_map.retain(|((vj_id, _), _)| vehicle_journeys.contains_key(&vj_id));
    attributes_map
}
``` 

or maybe even shorter
```rust
fn updated_stop_time_attributes(
    vehicle_journeys: &CollectionWithId<VehicleJourney>,
    attributes_map: &mut HashMap<(String, u32), String>,
) {
    attributes_map.retain(|((vj_id, _), _)| vehicle_journeys.contains_key(&vj_id));
}
```

File: src/ntfs/read.rs [hunk 5]
```diff
@@ -457,8 +463,8 @@ where
 }
 
 fn insert_stop_time_comment_link(
```
Comment (woshilapin) [inline src/ntfs/read.rs:465]: I feel that this function could be rewritten like this

```rust
fn insert_stop_time_comment_link(
    stop_time_comments: &mut HashMap<(String, u32), String>,
    stop_time_ids: &HashMap<&String, (String, u32)>,
    comments: &CollectionWithId<Comment>,
    comment_link: &CommentLink,
) -> Result<()> {
	if let Some(idx_sequence) = stop_time_ids.get(&comment_link.object_id) {
		if comments.contains_key(&comment_link.comment_id) {
			stop_time_comments.insert(idx_sequence.clone(), comment_link.comment_id.clone());
		} else {
			bail!(
				"comment.txt: comment_id={} not found",
				comment_link.comment_id
			)
		}
	} else {
		error!(
			"comment_links.txt: object_type={} object_id={} not found",
			comment_link.object_type.as_str(),
			comment_link.object_id
		);
	}
	Ok(())
}
```

File: src/add_prefix.rs (hunk 1 above)
Comment (woshilapin) [reply src/add_prefix.rs:57]: Hmm, the key is modified so it might not be possible.

File: src/model.rs (hunk 2 above)
Comment (woshilapin) [reply src/model.rs]: Oops, you forgot the parenthesis too?

File: src/ntfs/read.rs [hunk 6]
```diff
@@ -456,31 +462,56 @@ where
...
+//         Some(comment_idx) => comment_idx,
+//         None => bail!(
+//             "comment.txt: comment_id={} not found",
+//             comment_link.comment_id
+//         ),
+//     };
+//     stop_time_comments.insert(idx_sequence.clone(), comment_link.comment_id.clone());
+//     Ok(())
+// }
+
```
Comment (woshilapin) [inline src/ntfs/read.rs:492]: You can remove this comment I guess :wink:

File: src/add_prefix.rs [hunk 7]
```diff
@@ -52,6 +53,21 @@ where
...
+    prefix: &str,
+) -> HashMap<(String, u32), String> {
+    vehicle_journey_ids
+        .iter()
+        .map(|((trip_id, sequence), value)| {
+            (
+                (format!("{}{}", prefix, trip_id), *sequence),
```
Comment (ArnaudOggy) [inline src/add_prefix.rs:64]:
```suggestion
                (format!("{}:{}", prefix, trip_id), *sequence),
```
there's no : :thinking:  ?

Comment (patochectp) [reply src/add_prefix.rs:64]: No
impl AddPrefix for VehicleJourney {
    fn add_prefix(&mut self, prefix: &str) {
        self.id = prefix.to_string() + &self.id;
        self.route_id = prefix.to_string() + &self.route_id;
        self.dataset_id = prefix.to_string() + &self.dataset_id;
        self.company_id = prefix.to_string() + &self.company_id;
        self.service_id = prefix.to_string() + &self.service_id;
        self.trip_property_id = self
            .trip_property_id
            .as_ref()
            .map(|id| prefix.to_string() + id);
        self.geometry_id = self.geometry_id.as_ref().map(|id| prefix.to_string() + &id);
    }
}

Comment (ArnaudOggy) [reply src/add_prefix.rs:64]: Ok :+1:

**Pull request merged**